dashboard/src/components/users/EditUserForm/EditUserForm.tsx

```diff
@@ -0,0 +1,472 @@
+import ControlledSelect from '@/components/common/ControlledSelect';
+import { useDialog } from '@/components/common/DialogProvider';
+import Form from '@/components/common/Form';
+import { useGetRolesQuery } from '@/generated/graphql';
+import { useCurrentWorkspaceAndApplication } from '@/hooks/useCurrentWorkspaceAndApplication';
+import Button from '@/ui/v2/Button';
+import Chip from '@/ui/v2/Chip';
+import IconButton from '@/ui/v2/IconButton';
+import Input from '@/ui/v2/Input';
+import InputAdornment from '@/ui/v2/InputAdornment';
+import InputLabel from '@/ui/v2/InputLabel';
+import Option from '@/ui/v2/Option';
+import Select from '@/ui/v2/Select';
+import Text from '@/ui/v2/Text';
+import CopyIcon from '@/ui/v2/icons/CopyIcon';
+import { copy } from '@/utils/copy';
+import getUserRoles from '@/utils/settings/getUserRoles';
+import { yupResolver } from '@hookform/resolvers/yup';
+import { Avatar } from '@mui/material';
+import { format, formatDistance } from 'date-fns';
+import Image from 'next/image';
+import type { RemoteAppUser } from 'pages/[workspaceSlug]/[appSlug]/users';
+import { useEffect, useMemo } from 'react';
+import { FormProvider, useForm } from 'react-hook-form';
+import { twMerge } from 'tailwind-merge';
+import * as Yup from 'yup';
+
+export interface EditUserFormProps {
+  /**
+   * This is the selected user from the user's table.
+   */
+  user: RemoteAppUser;
+  /**
+   * Function to be called when the form is submitted.
+   */
+  onEditUser?: (
+    values: EditUserFormValues,
+    user: RemoteAppUser,
+  ) => Promise<void>;
+  /**
+   * Function to be called when the operation is cancelled.
+   */
+  onCancel?: VoidFunction;
+  /**
+   * Function to be called when banning the user.
+   */
+  onBanUser: (user: RemoteAppUser) => Promise<void>;
+  /**
+   * Function to be called when deleting the user.
+   */
+  onDeleteUser: (user: RemoteAppUser) => Promise<void>;
+}
+
+export const EditUserFormValidationSchema = Yup.object({
+  displayName: Yup.string(),
+  avatarURL: Yup.string(),
+  email: Yup.string()
+    .email('Invalid email address')
+    .required('This field is required.'),
+  emailVerified: Yup.boolean().optional(),
+  phoneNumber: Yup.string().nullable(),
+  phoneNumberVerified: Yup.boolean().optional(),
+  locale: Yup.string(),
+  defaultRole: Yup.string(),
+  roles: Yup.array().of(
+    Yup.object({
+      role: Yup.string(),
+      value: Yup.bool(),
+    }),
+  ),
+});
+
+export type EditUserFormValues = Yup.InferType<
+  typeof EditUserFormValidationSchema
+>;
+
+export default function EditUserForm({
+  user,
+  onEditUser,
+  onCancel,
+  onBanUser,
+  onDeleteUser,
+}: EditUserFormProps) {
+  const { onDirtyStateChange, openDialog } = useDialog();
+  const { currentApplication } = useCurrentWorkspaceAndApplication();
+
+  const { data } = useGetRolesQuery({
+    variables: { id: currentApplication.id },
+  });
+
+  const allAvailableProjectRoles = useMemo(
```

Comment:
Is `useMemo` really necessary here?

Reply:
Not even a bit, `getUserRoles` is pretty inexpensive to run and it doesn't depend from data outside itself (https://github.com/nhost/nhost/blob/4a57861354c1332a7043845d0308831565ba4993/dashboard/src/utils/settings/getUserRoles/getUserRoles.ts#L10) I'll actually take it out.